modules/tinymce/src/core/main/ts/newline/InsertBlock.ts

```diff
@@ -242,139 +224,198 @@ const addBrToBlockIfNeeded = function (dom, block) {
-  let tmpRng, container, offset, parentBlock;
-  let newBlock, fragment, containerBlock, parentBlockName, newBlockName, isAfterLastNodeInContainer;
-  const dom = editor.dom;
-  const schema = editor.schema, nonEmptyElementsMap = schema.getNonEmptyElements();
-  const rng = editor.selection.getRng();
 
-  // Creates a new block element by cloning the current one or creating a new one if the name is specified
-  // This function will also copy any text formatting from the parent block and add it to the new one
-  const createNewBlock = function (name?) {
-    let node = container, block, clonedNode, caretNode;
-    const textInlineElements = schema.getTextInlineElements();
-
-    if (name || parentBlockName === 'TABLE' || parentBlockName === 'HR') {
-      block = dom.create(name || newBlockName);
-    } else {
-      block = parentBlock.cloneNode(false);
-    }
+// Check that we aren't accidentally about to extract the entire parentNode.
+const handleEmptyRange = (parentNode: Node, rng: Range) => {
+  const testRange = document.createRange();
+  testRange.selectNode(parentNode);
+  if (testRange.compareBoundaryPoints(Range.START_TO_START, rng) !== 0 || testRange.compareBoundaryPoints(Range.END_TO_END, rng) !== 0) {
+    // We are not about to accidentally extract the entire parentNode, so we don't need to do anything special
+    return;
+  }
 
-    caretNode = block;
+  const newBlock = SugarElement.fromHtml('<br data-mce-bogus="1">').dom;
+  parentNode.insertBefore(newBlock, parentNode.firstChild);
+  rng.setStartAfter(newBlock);
+};
 
-    if (Settings.shouldKeepStyles(editor) === false) {
-      dom.setAttrib(block, 'style', null); // wipe out any styles that came over with the block
-      dom.setAttrib(block, 'class', null);
-    } else {
-      // Clone any parent styles
-      do {
-        if (textInlineElements[node.nodeName]) {
-          // Ignore caret or bookmark nodes when cloning
-          if (isCaretNode(node) || Bookmarks.isBookmarkNode(node)) {
-            continue;
-          }
-
-          clonedNode = node.cloneNode(false);
-          dom.setAttrib(clonedNode, 'id', ''); // Remove ID since it needs to be document unique
-
-          if (block.hasChildNodes()) {
-            clonedNode.appendChild(block.firstChild);
-            block.appendChild(clonedNode);
-          } else {
-            caretNode = clonedNode;
-            block.appendChild(clonedNode);
-          }
+// Used as an internal function below. Dragging it up here to make the function state explicit (and read-only).
+interface CreateNewBlock {
+  container: Node;
+  schema: Schema;
+  parentBlockName: string;
+  newBlockName: string;
+  dom: DOMUtils;
+  parentBlock: Element;
+  editor: Editor;
+  editableRoot: Node;
+}
+
+const _createNewBlock = ({ container, schema, parentBlockName, newBlockName, dom, parentBlock, editor, editableRoot }: CreateNewBlock, name: string | undefined): Node => {
```

Comment:
Hmm... this is a rather inconsistent naming schema, as using a underscore normally means the variable isn't used in a function argument at least. Can we come up with a name that doesn't do this? One thing we've done in a few places is use "internal" in the name.

Also why change this to an object and destructure it? I'd prefer this be kept as individual arguments as it'll minify better and also show if we aren't using variables better. If not, then please at least destructure it below instead of inside the function arguments as it's more readable.

Reply:
Another thing, if you're passing `editor` you probably don't need to pass `dom` or `schema`.

Reply:
Another option for the original point, is to not use a single object and use more focused objects specific based on what the data entails. eg: `parent: { name: string; block: Element }` as opposed to two variables for `parentBlockName` and `parentBlock`.

Reply:
- Renamed to `internalCreateNewBlock`
- I'm using objects and destructuring them so that you can't accidentally put arguments in the wrong order. Each of these has 5+ arguments (which is something I'd love to change as part of a larger rewrite) so using positional arguments feels like a lot of risk
- I've changed it to destructure below to make the function declaration easier to read
- Minification I can't really do anything about. Do we know how much of a difference this would make to total size / what the limitations of the minifier are around this?
- I tested unused parameters both where the internal functions are called (by the wrappers) and in the destructuring, and in both cases tsc complained loudly
- I have removed dom and schema from the arguments
- I can make that change, but I was trying to do as little as possible to change the actual code itself (otherwise where do you stop)